File: backend/src/main/java/kr/ygh/maple/common/feign/NexonError.java
package kr.ygh.maple.common.feign;

import java.util.Arrays;
import java.util.NoSuchElementException;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;

@Getter
@RequiredArgsConstructor
public enum NexonError {
    INTERNAL_SERVER_ERROR("OPENAPI00001", HttpStatus.BAD_GATEWAY, NexonError.PROBLEM_OCCURRED),
    UNAUTHORIZED_ACCESS("OPENAPI00002", HttpStatus.INTERNAL_SERVER_ERROR, NexonError.PROBLEM_OCCURRED),
    INVALID_IDENTIFIER("OPENAPI00003", HttpStatus.INTERNAL_SERVER_ERROR, NexonError.PROBLEM_OCCURRED),
    INVALID_PARAMETER("OPENAPI00004", HttpStatus.BAD_REQUEST, "접속한지 오래 되었거나 유요하지 않은 요청입니다."),
    INVALID_API_KEY("OPENAPI00005", HttpStatus.INTERNAL_SERVER_ERROR, NexonError.PROBLEM_OCCURRED),
    INVALID_GAME_OR_PATH("OPENAPI00006", HttpStatus.INTERNAL_SERVER_ERROR, NexonError.PROBLEM_OCCURRED),
    CALL_LIMIT_EXCEEDED("OPENAPI00007", HttpStatus.TOO_MANY_REQUESTS, "호출 횟수가 많습니다. 잠시 후 다시 시도해주세요."),
    DATA_BEING_PREPARED("OPENAPI00009", HttpStatus.BAD_GATEWAY, "넥슨 API 서비스에서 데이터 준비 중입니다."),
    SERVICE_UNDER_MAINTENANCE("OPENAPI00010", HttpStatus.BAD_GATEWAY, "넥슨 서비스가 점검 중입니다."),
    API_UNDER_MAINTENANCE("OPENAPI00011", HttpStatus.BAD_GATEWAY, "넥슨 API 서비스가 점검 중입니다.");

    private static final String PROBLEM_OCCURRED = "서버에서 넥슨 API 호출 중 문제가 발생하였습니다.";
    private final String code;
    private final HttpStatus responseStatus;
    private final String responseMessage;

    public static NexonError from(String codeIncluded) {
        return Arrays.stream(values())
                .filter(error -> codeIncluded.contains(error.code))
                .findAny()
                .orElseThrow(NoSuchElementException::new);
    }
}

File: proxy/src/nexon-error.ts
export class NexonApiError extends Error {
  constructor(
    public readonly status: number,
    message: string,
  ) {
    super(message);
    this.name = "NexonApiError";
  }
}

const KNOWN_ERRORS: Array<{ code: string; status: number; message: string }> = [
  { code: "OPENAPI00001", status: 502, message: "넥슨 API 에서 오류가 발생하였습니다." },
  { code: "OPENAPI00002", status: 500, message: "넥슨 API 권한이 없습니다." },
  { code: "OPENAPI00003", status: 500, message: "접속한지 오래 되었거나 유효하지 않은 식별자입니다." },
  { code: "OPENAPI00004", status: 400, message: "캐릭터가 존재하지 않거나 유효하지 않은 요청입니다." },
  { code: "OPENAPI00005", status: 500, message: "유효하지 않은 API 키입니다." },
  { code: "OPENAPI00006", status: 500, message: "유효하지 않은 게임 또는 API 경로입니다." },
  { code: "OPENAPI00007", status: 429, message: "호출 횟수가 많습니다. 잠시 후 다시 시도해주세요." },
  { code: "OPENAPI00009", status: 502, message: "넥슨 API 데이터 준비 중입니다." },
  { code: "OPENAPI00010", status: 502, message: "게임 점검 중입니다." },
  { code: "OPENAPI00011", status: 502, message: "넥슨 API 점검 중입니다." },
];

export async function toNexonApiError(response: Response): Promise<NexonApiError> {
  const body = await response.text().catch(() => "");
  const matched = KNOWN_ERRORS.find((error) => body.includes(error.code));

  if (matched) return new NexonApiError(matched.status, matched.message);

  return new NexonApiError(500, "서버에서 넥슨 API 호출 중 알 수 없는 문제가 발생하였습니다.");
}
